src/main/scala/nl/biopet/tools/bamstats/generate/Generate.scala

```diff
@@ -56,257 +54,134 @@ object Generate extends ToolCommand[Args] {
-    val groups =
-      contigs.foldLeft((List[List[BedRecord]](), List[BedRecord](), 0L)) {
-        case ((finalList, tempList, oldSize), b) =>
-          if (oldSize < threadBinSize)
-            (finalList, b :: tempList, oldSize + b.length)
-          else (tempList :: finalList, b :: Nil, b.length)
-      }
-    val contigsFutures =
-      (groups._2 :: groups._1).map(x => processThread(x, bamFile))
-
-    val unmappedStats = processUnmappedReads(bamFile)
-    val (stats, contigStats) = waitOnFutures(contigsFutures)
-    stats += Await.result(unmappedStats, Duration.Inf)
-
-    if (tsvOutput) {
-      stats.flagstat.writeAsTsv(
-        new File(outputDir, "flagstats.tsv")
-      )
-
-      stats.insertSizeHistogram.writeFilesAndPlot(outputDir,
-                                                  "insertsize",
-                                                  "Insertsize",
-                                                  "Reads",
-                                                  "Insertsize distribution")
-
-      stats.mappingQualityHistogram.writeFilesAndPlot(
-        outputDir,
-        "mappingQuality",
-        "Mapping Quality",
-        "Reads",
-        "Mapping Quality distribution")
-
-      stats.clippingHistogram.writeFilesAndPlot(outputDir,
-                                                "clipping",
-                                                "CLipped bases",
-                                                "Reads",
-                                                "Clipping distribution")
 
-      stats.leftClippingHistogram.writeFilesAndPlot(
-        outputDir,
-        "left_clipping",
-        "CLipped bases",
-        "Reads",
-        "Left Clipping distribution")
-
-      stats.rightClippingHistogram.writeFilesAndPlot(
-        outputDir,
-        "right_clipping",
-        "CLipped bases",
-        "Reads",
-        "Right Clipping distribution")
-
-      stats._3_ClippingHistogram.writeFilesAndPlot(
-        outputDir,
-        "3prime_clipping",
-        "CLipped bases",
-        "Reads",
-        "3 Prime Clipping distribution")
+    val stats = extractStats(
+      bamFile = cmdArgs.bamFile,
+      bedFile = cmdArgs.bedFile,
+      sequenceDict = sequenceDict,
+      excludePartialReads = cmdArgs.excludePartialReads,
+      getUnmappedReads = cmdArgs.getUnmappedReads
+    )
 
-      stats._5_ClippingHistogram.writeFilesAndPlot(
-        outputDir,
-        "5prime_clipping",
-        "CLipped bases",
-        "Reads",
-        "5 Prime Clipping distribution")
+    if (cmdArgs.tsvOutputs) {
+      writeStatsToTsv(stats, outputDir = cmdArgs.outputDir)
     }
 
-    val statsWriter = new PrintWriter(new File(outputDir, "bamstats.json"))
-    val totalStats = stats.toSummaryMap
-    val statsMap = Map(
-      "total" -> totalStats,
-      "contigs" -> contigStats
-    )
-    statsWriter.println(Json.stringify(conversions.mapToJson(statsMap)))
+    val groupedStats = Root.fromGroupStats(
```

Comment:
Maybe we should rename `Root` to `BamStatsRoot`? (same for seqstat). Might be confusing when combining the 2 or more

Reply:
Agree